File: frontend/services/tagService.ts
/**
 * Tag Service - 标签服务
 * 处理标签相关的API调用
 */

import { http } from '@/lib/utils/http';

export interface Tag {
  id: number;
  name: string;
  slug: string;
  description?: string;
  color?: string;
  icon?: string;
  created_at: string;
  updated_at: string;
}

export interface TagStats {
  tag_id: number;
  posts_count: number;
  published_count: number;
  recent_count: number;
}

export interface TagWithWeight extends Tag {
  count: number;
  weight: number;
}

export interface TagListResponse {
  items: Tag[];
  total: number;
  skip: number;
  limit: number;
}

export interface CreateTagData {
  name: string;
  description?: string;
  color?: string;
  icon?: string;
}

export interface UpdateTagData {
  name?: string;
  description?: string;
  color?: string;
  icon?: string;
}

/**
 * 获取标签列表
 */
export async function getTags(params?: {
  skip?: number;
  limit?: number;
  search?: string;
  sort_by?: 'name' | 'posts_count' | 'created' | 'popular';
}): Promise<TagListResponse> {
  return http.get('/tags', { params });
}

/**
 * 获取热门标签
 */
export async function getPopularTags(params?: {
  limit?: number;
  days?: number;
}): Promise<Tag[]> {
  return http.get('/tags/popular', { params });
}

/**
 * 获取标签云
 */
export async function getTagCloud(params?: {
  min_posts?: number;
  limit?: number;
}): Promise<TagWithWeight[]> {
  return http.get('/tags/cloud', { params });
}

/**
 * 获取单个标签
 */
export async function getTag(tagId: number): Promise<Tag> {
  return http.get(`/tags/${tagId}`);
}

/**
 * 获取标签统计信息
 */
export async function getTagStats(tagId: number): Promise<TagStats> {
  return http.get(`/tags/${tagId}/stats`);
}

/**
 * 获取相关标签
 */
export async function getRelatedTags(tagId: number, params?: {
  limit?: number;
}): Promise<Tag[]> {
  return http.get(`/tags/${tagId}/related`, { params });
}

/**
 * 创建标签
 */
export async function createTag(data: CreateTagData): Promise<Tag> {
  return http.post('/tags', data);
}

/**
 * 更新标签
 */
export async function updateTag(tagId: number, data: UpdateTagData): Promise<Tag> {
  return http.put(`/tags/${tagId}`, data);
}

/**
 * 删除标签
 */
export async function deleteTag(tagId: number): Promise<void> {
  return http.delete(`/tags/${tagId}`);
}

/**
 * 合并标签
 */
export async function mergeTags(sourceId: number, targetId: number): Promise<{ message: string }> {
  return http.post('/tags/merge', null, {
    params: { source_id: sourceId, target_id: targetId }
  });
}

/**
 * 通过 slug 获取标签
 */
export async function getTagBySlug(slug: string): Promise<Tag> {
  return http.get(`/tags/slug/${slug}`);
}

/**
 * 搜索标签
 */
export async function searchTags(query: string, limit?: number): Promise<Tag[]> {
  return getTags({
    search: query,
    limit: limit || 20
  }).then(response => response.items);
}

/**
 * 获取多个标签的统计信息
 */
export async function getMultipleTagStats(tagIds: number[]): Promise<TagStats[]> {
  const promises = tagIds.map(id => getTagStats(id));
  return Promise.all(promises);
}

/**
 * 获取标签的建议颜色
 */
export function getSuggestedColor(tagName: string): string {
  const colors = [
    '#00f0ff', // cyan
    '#9d00ff', // purple
    '#ff0080', // pink
    '#00ff88', // green
    '#f0ff00', // yellow
    '#ff6b00', // orange
    '#0066ff', // blue
    '#ff0066', // red
  ];

  let hash = 0;
  for (let i = 0; i < tagName.length; i++) {
    hash = tagName.charCodeAt(i) + ((hash << 5) - hash);
  }

  return colors[Math.abs(hash) % colors.length];
}

/**
 * 批量创建标签
 */
export async function bulkCreateTags(names: string[]): Promise<Tag[]> {
  const promises = names.map(name =>
    createTag({
      name,
      color: getSuggestedColor(name)
    })
  );

  return Promise.all(promises);
}

/**
 * 标签服务类
 */
export class TagService {
  private baseUrl: string;

  constructor(baseUrl: string = '/tags') {
    this.baseUrl = baseUrl;
  }

  async getTags(params?: {
    skip?: number;
    limit?: number;
    search?: string;
    sort_by?: string;
  }): Promise<TagListResponse> {
    return http.get(this.baseUrl, { params });
  }

  async getPopularTags(limit: number = 20, days: number = 30): Promise<Tag[]> {
    return http.get(`${this.baseUrl}/popular`, { params: { limit, days } });
  }

  async getTagCloud(minPosts: number = 1, limit: number = 50): Promise<TagWithWeight[]> {
    return http.get(`${this.baseUrl}/cloud`, { params: { min_posts: minPosts, limit } });
  }

  async getTag(tagId: number): Promise<Tag> {
    return http.get(`${this.baseUrl}/${tagId}`);
  }

  async getTagStats(tagId: number): Promise<TagStats> {
    return http.get(`${this.baseUrl}/${tagId}/stats`);
  }

  async getRelatedTags(tagId: number, limit: number = 10): Promise<Tag[]> {
    return http.get(`${this.baseUrl}/${tagId}/related`, { params: { limit } });
  }

  async createTag(data: CreateTagData): Promise<Tag> {
    return http.post(this.baseUrl, data);
  }

  async updateTag(tagId: number, data: UpdateTagData): Promise<Tag> {
    return http.put(`${this.baseUrl}/${tagId}`, data);
  }

  async deleteTag(tagId: number): Promise<void> {
    return http.delete(`${this.baseUrl}/${tagId}`);
  }

  async mergeTags(sourceId: number, targetId: number): Promise<{ message: string }> {
    return http.post(`${this.baseUrl}/merge`, null, {
      params: { source_id: sourceId, target_id: targetId }
    });
  }

  // 批量操作
  async bulkDelete(tagIds: number[]): Promise<void> {
    const promises = tagIds.map(id => this.deleteTag(id));
    await Promise.all(promises);
  }

  async getTagWithStats(tagId: number): Promise<Tag & { stats: TagStats }> {
    const [tag, stats] = await Promise.all([
      this.getTag(tagId),
      this.getTagStats(tagId)
    ]);

    return { ...tag, stats };
  }

  // 缓存相关
  private cache = new Map<string, { data: any; expiry: number }>();

  async getCachedTags(cacheKey: string = 'tags', ttl: number = 60000): Promise<Tag[]> {
    const cached = this.cache.get(cacheKey);
    if (cached && cached.expiry > Date.now()) {
      return cached.data;
    }

    const response = await this.getTags({ limit: 100 });
    this.cache.set(cacheKey, {
      data: response.items,
      expiry: Date.now() + ttl
    });

    return response.items;
  }

  clearCache(cacheKey?: string): void {
    if (cacheKey) {
      this.cache.delete(cacheKey);
    } else {
      this.cache.clear();
    }
  }
}

// 导出默认实例
export const tagService = new TagService();

/**
 * React Hook for Tags
 */
import { useEffect, useState } from 'react';

export function useTags(params?: {
  skip?: number;
  limit?: number;
  search?: string;
  sort_by?: string;
}) {
  const [data, setData] = useState<TagListResponse | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    const fetchTags = async () => {
      try {
        setLoading(true);
        const response = await getTags(params);
        setData(response);
      } catch (err) {
        setError(err as Error);
      } finally {
        setLoading(false);
      }
    };

    fetchTags();
  }, [JSON.stringify(params)]);

  return { data, loading, error };
}

export function useTag(tagId: number) {
  const [data, setData] = useState<Tag | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    const fetchTag = async () => {
      try {
        setLoading(true);
        const response = await getTag(tagId);
        setData(response);
      } catch (err) {
        setError(err as Error);
      } finally {
        setLoading(false);
      }
    };

    fetchTag();
  }, [tagId]);

  return { data, loading, error };
}

export function usePopularTags(limit: number = 20, days: number = 30) {
  const [data, setData] = useState<Tag[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    const fetchTags = async () => {
      try {
        setLoading(true);
        const response = await getPopularTags({ limit, days });
        setData(response);
      } catch (err) {
        setError(err as Error);
      } finally {
        setLoading(false);
      }
    };

    fetchTags();
  }, [limit, days]);

  return { data, loading, error };
}

export function useTagCloud(minPosts: number = 1, limit: number = 50) {
  const [data, setData] = useState<TagWithWeight[]>([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    const fetchCloud = async () => {
      try {
        setLoading(true);
        const response = await getTagCloud({ min_posts: minPosts, limit });
        setData(response);
      } catch (err) {
        setError(err as Error);
      } finally {
        setLoading(false);
      }
    };

    fetchCloud();
  }, [minPosts, limit]);

  return { data, loading, error };
}

File: backend/app/services/tag_service.py
"""
Tag Service - 标签服务
处理标签相关的业务逻辑
"""
from typing import List, Optional, Tuple
from sqlalchemy.orm import Session
from sqlalchemy import func, desc
from datetime import datetime, timedelta

from app.models.tag import Tag
from app.models.post import Post
from app.models.post_tag import post_tags
from app.schemas.tag import TagCreate, TagUpdate, TagStats


class TagService:
    """标签服务类"""

    def __init__(self, db: Session):
        self.db = db

    def get_tag(self, tag_id: int) -> Optional[Tag]:
        """获取单个标签"""
        return self.db.query(Tag).filter(Tag.id == tag_id).first()

    def get_tag_by_slug(self, slug: str) -> Optional[Tag]:
        """通过slug获取标签"""
        return self.db.query(Tag).filter(Tag.slug == slug).first()

    def get_tags(
        self,
        skip: int = 0,
        limit: int = 100,
        search: Optional[str] = None,
        sort_by: str = "name"
    ) -> Tuple[List[Tag], int]:
        """获取标签列表"""
        query = self.db.query(Tag)

        # 搜索过滤
        if search:
            query = query.filter(
                (Tag.name.ilike(f"%{search}%")) |
                (Tag.description.ilike(f"%{search}%"))
            )

        # 排序
        if sort_by == "name":
            query = query.order_by(Tag.name)
        elif sort_by == "posts_count":
            query = query.outerjoin(Tag.posts).group_by(Tag.id).order_by(
                desc(func.count(Post.id))
            )
        elif sort_by == "created":
            query = query.order_by(desc(Tag.created_at))
        elif sort_by == "popular":
            # 按最近30天的文章数排序
            since = datetime.utcnow() - timedelta(days=30)
            query = query.join(Tag.posts).filter(
                Post.published_at >= since
            ).group_by(Tag.id).order_by(desc(func.count(Post.id)))

        # 计数
        total = query.count()

        # 分页
        tags = query.offset(skip).limit(limit).all()

        return tags, total

    def create_tag(self, tag_data: TagCreate) -> Tag:
        """创建标签"""
        # 检查标签名是否已存在
        existing = self.db.query(Tag).filter(Tag.name == tag_data.name).first()
        if existing:
            raise ValueError(f"Tag '{tag_data.name}' already exists")

        # 生成slug
        slug = self._generate_slug(tag_data.name)

        tag = Tag(
            name=tag_data.name,
            slug=slug,
            description=tag_data.description,
            color=tag_data.color,
            icon=tag_data.icon
        )

        self.db.add(tag)
        self.db.commit()
        self.db.refresh(tag)

        return tag

    def update_tag(self, tag_id: int, tag_data: TagUpdate) -> Optional[Tag]:
        """更新标签"""
        tag = self.get_tag(tag_id)
        if not tag:
            return None

        # 更新字段
        if tag_data.name is not None:
            # 检查新名称是否与其他标签冲突
            existing = self.db.query(Tag).filter(
                Tag.name == tag_data.name,
                Tag.id != tag_id
            ).first()
            if existing:
                raise ValueError(f"Tag '{tag_data.name}' already exists")
            tag.name = tag_data.name
            tag.slug = self._generate_slug(tag_data.name)

        if tag_data.description is not None:
            tag.description = tag_data.description

        if tag_data.color is not None:
            tag.color = tag_data.color

        if tag_data.icon is not None:
            tag.icon = tag_data.icon

        tag.updated_at = datetime.utcnow()

        self.db.commit()
        self.db.refresh(tag)

        return tag

    def delete_tag(self, tag_id: int) -> bool:
        """删除标签"""
        tag = self.get_tag(tag_id)
        if not tag:
            return False

        self.db.delete(tag)
        self.db.commit()

        return True

    def get_tag_stats(self, tag_id: int) -> Optional[TagStats]:
        """获取标签统计信息"""
        tag = self.get_tag(tag_id)
        if not tag:
            return None

        # 文章总数
        posts_count = self.db.query(func.count(Post.id)).join(
            post_tags
        ).filter(post_tags.c.tag_id == tag_id).scalar()

        # 已发布文章数
        published_count = self.db.query(func.count(Post.id)).join(
            post_tags
        ).filter(
            post_tags.c.tag_id == tag_id,
            Post.status == "published"
        ).scalar()

        # 最近30天文章数
        since = datetime.utcnow() - timedelta(days=30)
        recent_count = self.db.query(func.count(Post.id)).join(
            post_tags
        ).filter(
            post_tags.c.tag_id == tag_id,
            Post.published_at >= since
        ).scalar()

        return TagStats(
            tag_id=tag.id,
            posts_count=posts_count,
            published_count=published_count,
            recent_count=recent_count
        )

    def get_popular_tags(self, limit: int = 20, days: int = 30) -> List[Tag]:
        """获取热门标签"""
        since = datetime.utcnow() - timedelta(days=days)

        tags = self.db.query(Tag).join(
            post_tags
        ).join(
            Post
        ).filter(
            Post.status == "published",
            Post.published_at >= since
        ).group_by(
            Tag.id
        ).order_by(
            desc(func.count(Post.id))
        ).limit(limit).all()

        return tags

    def get_related_tags(self, tag_id: int, limit: int = 10) -> List[Tag]:
        """获取相关标签（基于共同文章）"""
        # 找到与该标签共同出现最多的其他标签
        subquery = self.db.query(
            post_tags.c.post_id
        ).filter(
            post_tags.c.tag_id == tag_id
        ).subquery()

        tags = self.db.query(Tag).join(
            post_tags
        ).filter(
            post_tags.c.post_id.in_(subquery),
            post_tags.c.tag_id != tag_id
        ).group_by(
            Tag.id
        ).order_by(
            desc(func.count(post_tags.c.post_id))
        ).limit(limit).all()

        return tags

    def merge_tags(self, source_id: int, target_id: int) -> bool:
        """合并标签（将source标签的文章转移到target标签）"""
        source_tag = self.get_tag(source_id)
        target_tag = self.get_tag(target_id)

        if not source_tag or not target_tag:
            return False

        # 更新所有使用source标签的文章关联
        self.db.execute(
            post_tags.update()
            .where(post_tags.c.tag_id == source_id)
            .values(tag_id=target_id)
        )

        # 删除source标签
        self.db.delete(source_tag)
        self.db.commit()

        return True

    def _generate_slug(self, name: str) -> str:
        """生成URL友好的slug"""
        # 简单的slug生成
        slug = name.lower().strip()
        # 替换空格和特殊字符
        slug = "-".join(c for c in slug if c.isalnum() or c in (" ", "-"))
        slug = slug.replace(" ", "-")

        # 确保slug唯一
        counter = 1
        original_slug = slug
        while self.db.query(Tag).filter(Tag.slug == slug).first():
            slug = f"{original_slug}-{counter}"
            counter += 1

        return slug

    def get_or_create_tag(self, name: str) -> Tag:
        """获取或创建标签（用于导入等场景）"""
        tag = self.db.query(Tag).filter(Tag.name == name).first()
        if tag:
            return tag

        slug = self._generate_slug(name)
        tag = Tag(name=name, slug=slug)
        self.db.add(tag)
        self.db.commit()
        self.db.refresh(tag)

        return tag

    def bulk_create_tags(self, names: List[str]) -> List[Tag]:
        """批量创建标签"""
        tags = []
        for name in names:
            try:
                tag = self.get_or_create_tag(name)
                tags.append(tag)
            except Exception:
                continue

        return tags

    def get_tag_cloud(self, min_posts: int = 1, limit: int = 50) -> List[dict]:
        """获取标签云（包含权重）"""
        tags = self.db.query(
            Tag.id,
            Tag.name,
            Tag.slug,
            func.count(Post.id).label("count")
        ).join(
            post_tags
        ).join(
            Post
        ).filter(
            Post.status == "published"
        ).group_by(
            Tag.id
        ).having(
            func.count(Post.id) >= min_posts
        ).order_by(
            desc("count")
        ).limit(limit).all()

        # 计算权重（1-10）
        if tags:
            max_count = max(tag.count for tag in tags)
            min_count = min(tag.count for tag in tags)

            result = []
            for tag in tags:
                if max_count == min_count:
                    weight = 5
                else:
                    weight = 1 + int((tag.count - min_count) / (max_count - min_count) * 9)

                result.append({
                    "id": tag.id,
                    "name": tag.name,
                    "slug": tag.slug,
                    "count": tag.count,
                    "weight": weight
                })

            return result

        return []
